packages/dds/tree/src/tree/sparseTree.ts

```diff
@@ -0,0 +1,167 @@
+        }
+        return this.parentPath;
+    }
+
+    /**
+     * Gets a child, adding a ref to it.
+     * Creates child (with 1 ref) if needed.
+     */
+    public getOrCreateChild(key: FieldKey, index: number, data: () => TData): SparseNode<TData> {
+        let field = this.children.get(key);
+        if (field === undefined) {
+            field = [];
+            this.children.set(key, field);
+        }
+        // TODO: should do more optimized search (ex: binary search).
+        let child = field.find((c) => c.parentIndex === index);
+        if (child === undefined) {
+            child = new SparseNode(key, index, this, data());
+            field.push(child);
+            // Keep list sorted by index.
+            field.sort((a, b) => a.parentIndex - b.parentIndex);
+        }
+        return child;
+    }
+
+    /**
+     * Gets a child if it exists.
+     * Does NOT add a ref.
+     */
+    public tryGetChild(key: FieldKey, index: number): SparseNode<TData> | undefined {
+        const field = this.children.get(key);
+        if (field === undefined) {
+            return undefined;
+        }
+        // TODO: should do more optimized search (ex: binary search or better) using index.
+        return field.find((c) => c.parentIndex === index);
+    }
+
+    /**
+     * Removes reference from this to `child`.
+     * Since PathNodes are doubly linked,
+     * the caller must ensure that the reference from child to parent is also removed (or the child is no longer used).
+     */
+    public removeChild(child: SparseNode<TData>): void {
+        const key = child.parentField;
+        const field = this.children.get(key);
+        // TODO: should do more optimized search (ex: binary search or better) using child.parentIndex()
+        // Note that this is the index in the list of child paths, not the index within the field
+        const childIndex = field?.indexOf(child) ?? -1;
+        assert(childIndex !== -1, 0x35c /* child must be parented to be removed */);
+        field?.splice(childIndex, 1);
+        if (field?.length === 0) {
+            this.afterEmptyField(key);
+        }
+    }
+
+    /**
+     * Call this after directly editing the child array for a field to be empty.
+     * Handles cleaning up unneeded data
+     * (like the field in the map, and possibly this entire SparseNode and its parents if they are no longer needed.)
+     */
+    public afterEmptyField(key: FieldKey): void {
+        this.children.delete(key);
+        if (this.children.size === 0) {
+            this.disposeThis();
+        }
+    }
+
+    /**
+     * Removes this from parent if alive, and sets this to disposed.
+     * Must only be called when this node is no longer needed (has no references and no children).
+     *
+     * Allowed when dangling (but not when disposed).
+     */
+    private disposeThis(): void {
+        this.parentPath?.removeChild(this);
+    }
+}
+
+export function getRelativeNode<TData>(
```

Comment:
Consider renaming this to make it more obvious that the "relative" is a descendant. Maybe rename `from` to `root` as well.

Reply:
I made some changes that should help. I don't want to use `root` for the first parameter because I don't want to make it seem as though it must be the root node.